src/properties/trees.c

```diff
@@ -254,27 +254,39 @@ static igraph_error_t igraph_i_is_tree_visitor(const igraph_t *graph, igraph_int
  * \example examples/simple/igraph_kary_tree.c
  */
 igraph_error_t igraph_is_tree(const igraph_t *graph, igraph_bool_t *res, igraph_integer_t *root, igraph_neimode_t mode) {
+    igraph_bool_t is_tree = 0;
     igraph_integer_t iroot = 0;
     igraph_integer_t visited_count;
     igraph_integer_t vcount, ecount;
 
     vcount = igraph_vcount(graph);
     ecount = igraph_ecount(graph);
 
+    /* For undirected graphs, we can return early if we know from the cache
+     * that the graph is weakly connected and is a forest */
```

Comment:
Not just for undirected graphs, but also when `mode == IGRAPH_ALL`. So below it should be `(!igraph_is_directed(graph) || mode == IGRAPH_ALL)`. Am I missing something?